plugins/woocommerce/client/blocks/assets/js/shared/redux-stores/canvas-mode/index.ts

```diff
@@ -0,0 +1,103 @@
+/**
+ * External dependencies
+ */
+import { createReduxStore, register, select, dispatch } from '@wordpress/data';
+
+/**
+ * Internal dependencies
+ */
+import { ACTION_SET_CANVAS_MODE, STORE_NAME } from './constants';
+import type { CanvasModeState } from './types';
+
+type StoreState = CanvasModeState;
+
+type Actions = {
+	type: typeof ACTION_SET_CANVAS_MODE;
+	isEditMode: boolean;
+};
+
+const DEFAULT_STATE: StoreState = {
+	isEditMode: false,
+};
+
+const actions = {
+	setCanvasMode( isEditMode: boolean ) {
+		return {
+			type: ACTION_SET_CANVAS_MODE,
+			isEditMode,
+		};
+	},
+};
+
+const selectors = {
+	isEditMode() {
+		// Always check the current URL for the most up-to-date state.
+		try {
+			const url = new URL( window.location.href );
+			const canvas = url.searchParams.get( 'canvas' );
+			return canvas === 'edit';
+		} catch ( error ) {
+			// If URL parsing fails, return default state.
+			return false;
+		}
+	},
+};
+
+if ( typeof window !== 'undefined' ) {
+	const handleUrlChange = async () => {
+		try {
+			const url = new URL( window.location.href );
+			const canvas = url.searchParams.get( 'canvas' );
+			const isEditMode = canvas === 'edit';
+
+			// Update store state to trigger any subscribed components.
+			const storeDispatch = dispatch( STORE_NAME ) as {
+				setCanvasMode: ( isEditMode: boolean ) => void;
+			};
+
+			await Promise.resolve();
+
+			// Update store state - this triggers re-renders.
+			storeDispatch.setCanvasMode( isEditMode );
+		} catch ( error ) {
+			// If URL parsing fails, set to default state.
+			const storeDispatch = dispatch( STORE_NAME ) as {
+				setCanvasMode: ( isEditMode: boolean ) => void;
+			};
+
+			await Promise.resolve();
+			storeDispatch.setCanvasMode( false );
+		}
+	};
+
+	window.addEventListener( 'popstate', () => handleUrlChange() );
+	window.addEventListener( 'pushstate', () => handleUrlChange() );
```

Comment:
I did some tests and I noticed that the real issue is here.

In some cases, the URL isn't the updated one:


https://github.com/user-attachments/assets/50179f3f-c4d3-4738-9ea8-b57ba6fc6973

Reply:
Also I noticed that `dispatch(store).setCanvasMode` returns a Promise. So we should add await:

```ts
await dispatch(store).setCanvasMode
```